### PR TITLE
rgw/multisite: fix marker_tracker flush() call in incremental sync

### PR DESCRIPTION
**Pull request by smanjara**
https://tracker.ceph.com/issues/63378

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by smanjara** [comment]
jenkins test make check

File: src/rgw/driver/rados/rgw_data_sync.cc [hunk 1]
```diff
@@ -1852,12 +1852,11 @@ class RGWDataFullSyncShardCR : public RGWDataBaseSyncShardCR {
       omapvals.reset();
 
       drain_all();
+      yield call(marker_tracker->flush());
```
Comment (cbodley) [inline src/rgw/driver/rados/rgw_data_sync.cc:1855]: this adds a `flush()` in 2 cases where it wasn't before:

when `lost_lock` is true, that means some other radosgw has acquired it, read the sync status, and started sync. if we issue a racing write to the sync status here, that'll cause the other radosgw to fail with `ECANCELED` when it wants to record its own progress

when `lost_lock` is false, we're overwriting the sync status anyway to switch to `IncrementalSync`, so the `flush()` is wasted there

File: src/rgw/driver/rados/rgw_data_sync.cc [hunk 2]
```diff
@@ -2122,7 +2121,7 @@ class RGWDataIncSyncShardCR : public RGWDataBaseSyncShardCR {
       } while (true);
 
       drain_all();
-      yield marker_tracker->flush();
+      yield call(marker_tracker->flush());
```
Comment (cbodley) [inline src/rgw/driver/rados/rgw_data_sync.cc:2124]: yikes, good catch. without the `call()`, nothing would actually run that coroutine, and its memory would just leak

File: src/rgw/driver/rados/rgw_data_sync.cc (hunk 1 above)
Comment (smanjara) [reply src/rgw/driver/rados/rgw_data_sync.cc:1855]: hmm but then how do we handle the case of `lost_lock`. don't flush at all?
in incremental sync, we do seem to call flush() in all cases.

Comment (cbodley) [reply src/rgw/driver/rados/rgw_data_sync.cc:1855]: > how do we handle the case of lost_lock. don't flush at all?

right. the `drain_all()` means that all the 'sync single entry' coroutines have completed, so nothing else would try to flush. if another radosgw has the lock, it should be responsible for managing the sync status

> in incremental sync, we do seem to call flush() in all cases

i feel like incremental sync should avoid the `flush()` on `lost_lock` for the same reason

Comment (smanjara) [reply src/rgw/driver/rados/rgw_data_sync.cc:1855]: great, thanks for elaborating!

**Review by cbodley — APPROVED** [review]

**Comment by cbodley** [comment]
this fix looks correct, but i still saw a segfault in [multisite tests](http://qa-proxy.ceph.com/teuthology/cbodley-2023-11-23_19:53:09-rgw-wip-cbodley-testing-distro-default-smithi/7465212/teuthology.log):
```
2023-11-23T21:56:12.727 INFO:tasks.rgw.c2.client.0.smithi114.stdout:*** Caught signal (Segmentation fault) **
2023-11-23T21:56:12.727 INFO:tasks.rgw.c2.client.0.smithi114.stdout: in thread 7f44ee58f640 thread_name:data-sync
2023-11-23T21:56:12.728 INFO:tasks.rgw.c2.client.0.smithi114.stdout: ceph version 18.0.0-7493-g80625caf (80625caf87ea00941730e4beea11d67e7101d49e) reef (dev)
2023-11-23T21:56:12.728 INFO:tasks.rgw.c2.client.0.smithi114.stdout: 1: /lib64/libc.so.6(+0x54db0) [0x7f4515854db0]
2023-11-23T21:56:12.728 INFO:tasks.rgw.c2.client.0.smithi114.stdout: 2: /lib64/libc.so.6(+0xc1992) [0x7f45158c1992]
2023-11-23T21:56:12.728 INFO:tasks.rgw.c2.client.0.smithi114.stdout: 3: radosgw(+0xc7dc4c) [0x55d52979ac4c]
2023-11-23T21:56:12.729 INFO:tasks.rgw.c2.client.0.smithi114.stdout: 4: radosgw(+0x383692) [0x55d528ea0692]
2023-11-23T21:56:12.729 INFO:tasks.rgw.c2.client.0.smithi114.stdout: 5: radosgw(+0x832359) [0x55d52934f359]
2023-11-23T21:56:12.729 INFO:tasks.rgw.c2.client.0.smithi114.stdout: 6: radosgw(+0x9c6b17) [0x55d5294e3b17]
2023-11-23T21:56:12.729 INFO:tasks.rgw.c2.client.0.smithi114.stdout: 7: (RGWCoroutinesStack::operate(DoutPrefixProvider const*, RGWCoroutinesEnv*)+0x125) [0x55d5290c7a65]
2023-11-23T21:56:12.730 INFO:tasks.rgw.c2.client.0.smithi114.stdout: 8: (RGWCoroutinesManager::run(DoutPrefixProvider const*, std::__cxx11::list<RGWCoroutinesStack*, std::allocator<RGWCoroutinesStack*> >&)+0x489) [0x55d5290c94a9]
2023-11-23T21:56:12.730 INFO:tasks.rgw.c2.client.0.smithi114.stdout: 9: (RGWCoroutinesManager::run(DoutPrefixProvider const*, RGWCoroutine*)+0xad) [0x55d5290ca29d]
2023-11-23T21:56:12.730 INFO:tasks.rgw.c2.client.0.smithi114.stdout: 10: (RGWRemoteDataLog::run_sync(DoutPrefixProvider const*, int)+0x4dc) [0x55d5294ee54c]
2023-11-23T21:56:12.730 INFO:tasks.rgw.c2.client.0.smithi114.stdout: 11: radosgw(+0x770878) [0x55d52928d878]
2023-11-23T21:56:12.730 INFO:tasks.rgw.c2.client.0.smithi114.stdout: 12: (RGWRadosThread::Worker::entry()+0xb3) [0x55d529292e03]
2023-11-23T21:56:12.730 INFO:tasks.rgw.c2.client.0.smithi114.stdout: 13: /lib64/libc.so.6(+0x9f802) [0x7f451589f802]
2023-11-23T21:56:12.731 INFO:tasks.rgw.c2.client.0.smithi114.stdout: 14: /lib64/libc.so.6(+0x3f450) [0x7f451583f450]
```

**Comment by smanjara** [comment]
> this fix looks correct, but i still saw a segfault in [multisite tests](http://qa-proxy.ceph.com/teuthology/cbodley-2023-11-23_19:53:09-rgw-wip-cbodley-testing-distro-default-smithi/7465212/teuthology.log):
> 

thanks, i'll take a look.

**Comment by smanjara** [comment]
I see that we call `rgw_rados_ref` directly now. something changed as part of https://github.com/ceph/ceph/pull/50359 ?

**Comment by cbodley** [comment]
> I see that we call `rgw_rados_ref` directly now. something changed as part of #50359 ?

yeah. in https://github.com/ceph/ceph/pull/50359#discussion_r1416323183 i point out something that looks suspicious

**Pull request merged**